proposal/custom-url-resolution.md

```diff
@@ -0,0 +1,247 @@
+
+## Summary
+
+> This section is non-normative.
+
+This proposal defines a standardized way to remap the sass-url references to the final location on the server or inline reference.
+
+This is accomplished by running sass-url references through the url rewriting plugin if a url rewriting plugin has been defined.
+
+### Example
+
+This example demonstrates the possibility of rewriting urls to inline assets using data URIs.
+
+Input:
+
+```scss
+.upload-icon {
+  background-url: sass-url('../icons/upload-icon.svg');
+}
+```
+
+Output:
+
+```css
+.upload-icon {
+  background-url: url('data:image/svg+xml;base64,Jggg==');
+}
+```
+
+## Function
+
+This proposal introduces logic for a new function: `sass-url("...")`, this function allow users to define a url that gets rewritten based on the config to ensure it can be loaded correctly from the browser.
+
+### Steps
+
+Whenever a sass-url reference is encountered in a sass file the following steps should be executed:
+
+- A sass-url reference is encountered in the sass file
+- From the url function we extract the parameter value
+- Rewrite the parameter value if it contains any variables ([see using variables](#using-variables))
+- Pass the parameter value to the url rewrite plugin along with the canonical url of the current sass file
+- This plugin than returns a string or calls the done callback with a string value, if it returns anything else or nothing it should throw an error ([see JavaScript API](#javascript-api))
+- The new url value that we received from the url rewrite plugin than gets used to replace the original url value
+
+_Note: if there is no urlRewrite plugin/function, none of these steps should be executed._
+
+### JavaScript API
+
+At the core of remapping the sass-url references is the JavaScript API which allows users to return a new sass-url reference, based on the canonical url of the sass file and original sass-url reference.
+
+The first parameter of the urlRewrite function is an object with the following values:
+
+- `file`: The canonical url of the sass file that references sass-url.
+- `url`: The sass-url reference, for example with `sass-url(file://../assets/test.png)` it would be `file://../assets/test.png`.
+
+The second parameter of the urlRewrite function is an optional done callback that is used when performing asynchronous operations.
+
+This callback function takes in two parameters:
+
+- `error`: This can be null or an Error object, `null` means there is no error.
+- `url`: This is the `string` that gets used to replace the sass-url reference. This cannot be null or anything else, otherwise sass will throw an error.
+
+Asynchronous example:
+
+```TypeScript
+let sassOptions = {
+  // Rewrite urls asynchronously
+  rewriteUrl: async (urlReference: { file: string, url: string }, done: (error: Error | null, url: string | null) => void): void => {
```

Comment:
Passing a `done` callback does not make sense to me. The result should be returned in the promise instead (an async function already returns a promise anyway)

Reply:
Was just mimicing existing sass api's but will update